[1.14]adding/removing JigsawPieces to Village Pools

Cheaterpaul opened the pull request:
(while there is no option to add buildings to a village)

2 events are fired:
- **JigsawPatternInitEvent** - fired in constructor of JigsawPattern. There you can add/remove JigsawPieces
- **StructureJigsawPoolInitEvent** - fired when a jigsaw based structure inititizes its JigsawPatterns. There you can add your own JigsawPattern before the structure need them.

Cheaterpaul committed:
adding feature to add JigsawPieces to Village Pools

Cheaterpaul committed:
some modifications

Gegy commented on patches/minecraft/net/minecraft/world/gen/feature/jigsaw/JigsawPattern.java.patch:
```diff
@@ -0,0 +1,35 @@
+       return this.field_214954_f;
+    }
+ 
++   public void addBuilding(Pair<JigsawPiece,Integer> piece)
```
Maybe rather construct the pair internally for the list and pass the values as separate parameters

Wyn-Price commented on patches/minecraft/net/minecraft/world/gen/feature/structure/VillagePieces.java.patch:
```diff
@@ -0,0 +1,31 @@
++import net.minecraft.world.gen.feature.jigsaw.JigsawManager;
++import net.minecraft.world.gen.feature.jigsaw.JigsawPiece;
++import net.minecraft.world.gen.feature.template.TemplateManager;
++import net.minecraftforge.common.MinecraftForge;
```
You shouldn't be messing with the imports in a patch file. Use fully qualified names instead (`net.minecraftforge.common.MinecraftForge`).

MajorTuvok commented on src/test/java/net/minecraftforge/debug/village/VillageJigsawPiecesInitTest.java:
```diff
@@ -0,0 +1,58 @@
+
+    public VillageJigsawPiecesInitTest()
+    {
+        MinecraftForge.EVENT_BUS.register(this);
```
No need to reigster again - there's no instance Method which would recieve an event. (The static Method is handeled by the EventBusSubscriber annotation)

MajorTuvok commented on src/main/java/net/minecraftforge/event/VillageStructureInitEvent.java:
```diff
@@ -0,0 +1,37 @@
+ * get {@link net.minecraft.world.gen.feature.jigsaw.JigsawPatternRegistry#get(ResourceLocation)} from {@link net.minecraft.world.gen.feature.jigsaw.JigsawManager#field_214891_a}
+ * and call {@link net.minecraft.world.gen.feature.jigsaw.JigsawPattern#addBuilding(Pair)
+ */
+public class VillageStructureInitEvent extends Event {
```
What about adding an ImmutableList.Builder and a ResourceLocation to this Event and allowing people to addBuildings here instead of getting the JigsawPattern from the Registry to add their own. Firing this Event then from the constructor of JigsawPattern would reduce the patch size a bit and give this event some functionality at the cost of a bit of performance in the constructor... I don't think it is invoked too often, right?

MajorTuvok reviewed (commented):
Some comments about perhaps giving the Event some functionality instead of having an empty Event lying around (which would also reduce patch size)

Cheaterpaul commented:
~~I thought about deleting or replacing, and until now i came to the idea of firing an event for every building that is added (big patch size, not neary clean) or~~ have a way of identifying methods or attributes for all JigsawPieces like SingleJigsawPiece to remove them out of the list.

edit:
Minecraft's way of adding pools / buildings is too disordered for a clean method

Cheaterpaul committed:
rearranging stuff

Cheaterpaul committed:
now able to remove and JigsawPattern for better compatibility,
fixed some issues,
updated test

Cheaterpaul committed:
fixed patch imports

LexManos commented on patches/minecraft/net/minecraft/world/gen/feature/jigsaw/EmptyJigsawPiece.java.patch:
```diff
@@ -0,0 +1,19 @@
+ import java.util.Collections;
+ import java.util.List;
+ import java.util.Random;
++import net.minecraft.util.ResourceLocation;
```
Your patches are never supposed to include imports

LexManos commented on patches/minecraft/net/minecraft/world/gen/feature/jigsaw/EmptyJigsawPiece.java.patch:
```diff
@@ -0,0 +1,19 @@
+    }
++
++   public ResourceLocation getIdentifier() {
++      return new ResourceLocation("empty");
```
This should be cached for performance take.

LexManos commented on patches/minecraft/net/minecraft/world/gen/feature/jigsaw/FeatureJigsawPiece.java.patch:
```diff
@@ -0,0 +1,19 @@
+    }
++
++   public ResourceLocation getIdentifier() {
++      return Registry.field_218379_q.func_177774_c(this.field_214870_a.field_222737_a);
```
If its a Forge registry it has getRegistryName, but either way this should be cached.

LexManos commented on patches/minecraft/net/minecraft/world/gen/feature/jigsaw/JigsawPattern.java.patch:
```diff
@@ -0,0 +1,67 @@
++      field_214953_e.clear();
++      for(Pair<JigsawPiece, Integer> e: field_214952_d){
++         for(Integer integer = 0; integer < e.getSecond(); integer = integer + 1) {
++            this.field_214953_e.add(e.getFirst().func_214845_a(field_214955_g));
```
don't copy paste code with bad names. this could a lot smaller patch if written by a human

LexManos commented on patches/minecraft/net/minecraft/world/gen/feature/jigsaw/JigsawPattern.java.patch:
```diff
@@ -0,0 +1,67 @@
++    */
++   @Deprecated
++   public void addBuildings(List<Pair<JigsawPiece,Integer>> pieces) {
++      this.field_214952_d.addAll(pieces);
```
Why add a directly deperacted method?

LexManos commented on patches/minecraft/net/minecraft/world/gen/feature/jigsaw/JigsawPattern.java.patch:
```diff
@@ -0,0 +1,67 @@
++
++   /**
++    * use {@link net.minecraftforge.event.world.jigsaw.JigsawStructurePoolInitEvent} for removing buildings
++    */
```
If you're telling people to use the event, then dont expose this.

Cheaterpaul replied:
how else should this work? just comment "see {@link net.minecraftforge.event.world.JigsawStructurePoolInitEvent}".
if not public the JigsawPiece List must be public?

Cheaterpaul replied:
if the fields are public, we dont need addBuilding/buildList/removeBuilding in JigsawPattern

MajorTuvok replied:
You could fire the event at construction time of the JigsawPattern. Yes that'd be more events, but firing one event per Pattern shouldn't be too bad, especially as that's done only once. This way you wouldn't expose anything, you wouldn't need to iterate over stuff after the event, etc. You'd just pass the freshly created arraylist into the event, allow people to modify that and then apply that to the field. (If you feel like preventing people from holding references to the list, you could for example copy it or just provide add and remove Methods in the event instead of giving people the list, etc.) 

(That'd be one solution which wouldn't make stuff public that shouldn't be public - there are certainly more)

LexManos commented on patches/minecraft/net/minecraft/world/gen/feature/jigsaw/JigsawPiece.java.patch:
```diff
@@ -0,0 +1,18 @@
+       return 1;
+    }
++
++   public abstract ResourceLocation getIdentifier();
```
Adding abstract methods is a binary break, as other mod systems wont implement this and get abstract method exception, so you need to find a sane default.

LexManos commented on patches/minecraft/net/minecraft/world/gen/feature/jigsaw/ListJigsawPiece.java.patch:
```diff
@@ -0,0 +1,19 @@
+    }
++
++   public ResourceLocation getIdentifier() {
++      return new ResourceLocation(this.field_214867_a.stream().map(Object::toString).collect(Collectors.joining(", ")));
```
What type of name is this... seems like it'd change if anything in the list changes... which is bad for reference sake

LexManos commented on src/main/java/net/minecraftforge/event/ForgeEventFactory.java:
```diff
@@ -699,4 +701,13 @@ public static boolean onPistonMovePost(World world, BlockPos pos, Direction dire
+    public static void onJigsawStructurePiecesInit(JigsawStructurePoolInitEvent event)
+    {
+        MinecraftForge.EVENT_BUS.post(event);
+        for (Map.Entry<ResourceLocation, ImmutableList<ResourceLocation>> entry: event.getRemoveBuildings().entrySet())
```
Pass the fields needed into this event so you dont have to have a public remove/add

Cheaterpaul replied:
should this mean to add all JigsawLists (JigsawPattern#field_214952_d) into the event? or just compute the removal & addition inthe event and not from the ForgeEventFactory?

LexManos commented on src/main/java/net/minecraftforge/event/world/jigsaw/ForgeVillagePools.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    static {
+        ForgeEventFactory.onJigsawStructurePiecesInit(new JigsawStructurePoolInitEvent.Village());
```
These special classes are a definite no, what was the logic behind this and the hacky static init event firing... fire the events normally at a specific time.

Cheaterpaul replied:
because vanilla uses the same shitty way to add the Patters/Pieces only on time at runtime

LexManos replied:
Just because vanilla does something dumb, doesnt mean we have to

MajorTuvok replied:
You already call init, so why don't you use that instead of this static initialiser?!?

LexManos commented on src/main/java/net/minecraftforge/event/world/jigsaw/JigsawStructurePoolInitEvent.java:
```diff
@@ -0,0 +1,155 @@
+        if(newBuildings.containsKey(type))
+            newBuildings.get(type).addAll(pieces);
+        else
+            newBuildings.put(type,new ImmutableList.Builder<Pair<JigsawPiece, Integer>>().addAll(pieces));
```
map.computeIfAbsent()

tterrag1098 commented on patches/minecraft/net/minecraft/world/gen/feature/jigsaw/JigsawPatternRegistry.java.patch:
```diff
@@ -0,0 +1,12 @@
+    }
+ 
++   /**
++    * use {@link net.minecraftforge.event.world.jigsaw.JigsawStructurePoolInitEvent#registerNewJigsawPattern(JigsawPattern)} for registering new patterns for existing structures
```
Do not add javadoc to vanilla methods

JoJoDeveloping replied:
Use the bot instead

tterrag1098 replied:
References to forge code/patches should definitely not be in MCP comments...

MajorTuvok commented on src/main/java/net/minecraftforge/event/world/jigsaw/ForgePillageOutpostPools.java:
```diff
@@ -0,0 +1,32 @@
+public class ForgePillageOutpostPools {
+
+    public static void init() {
+    }
```
You even call init at the apropiate time, so why does this class have a static initialiser?!?

Cheaterpaul replied:
^^to fire the event only on time, because vanilla does the same. But I found a solution

MajorTuvok replied:
Your comment makes no sense... Though you did what we wanted (just without this init Method at all)

Cheaterpaul commented:
^^to fire the event only on time, because vanilla does the same. But I found a solution

Cheaterpaul committed:
requested changes

Cheaterpaul committed:
fix

Cheaterpaul committed:
moved removeBuilding/addBuilding,
make JigsawPiece lists public

Cheaterpaul committed:
now fires two events:
one on jigsawpattern constructor to directly add/remove all pices,
one on Village/PillageOutpost to register jigsawpattern before the structure is constructed

Cheaterpaul commented:
> 
> 
> You could fire the event at construction time of the JigsawPattern. Yes that'd be more events, but firing one event per Pattern shouldn't be too bad, especially as that's done only once. This way you wouldn't expose anything, you wouldn't need to iterate over stuff after the event, etc. You'd just pass the freshly created arraylist into the event, allow people to modify that and then apply that to the field. (If you feel like preventing people from holding references to the list, you could for example copy it or just provide add and remove Methods in the event instead of giving people the list, etc.)
> 
> (That'd be one solution which wouldn't make stuff public that shouldn't be public - there are certainly more)

Cheaterpaul committed:
fix merge conflicts

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=6142) <br/>All committers have signed the CLA.

Cheaterpaul committed:
edit comment

Cheaterpaul commented on src/main/java/net/minecraftforge/event/world/JigsawPatternInitEvent.java:
```diff
@@ -0,0 +1,129 @@
+
+        public abstract ResourceLocation getStructureRegistryName();
+
+        public static class Village extends StructureJigsawPoolInitEvent{
```
```suggestion
        public static class Village extends StructureJigsawPoolInitEvent{
            private static boolean fired;

            public Village() {
                fired = true;
            }

            public static boolean isFired(){
                return fired;
            }

            @Override
            public ResourceLocation getStructureRegistryName() {
                return Structures.VILLAGE.getRegistryName();
            }
        }
```

Cheaterpaul commented on patches/minecraft/net/minecraft/world/gen/feature/structure/VillagePieces.java.patch:
```diff
@@ -0,0 +1,10 @@
+       SavannaVillagePools.func_214745_a();
+       DesertVillagePools.func_222739_a();
+       TaigaVillagePools.func_214806_a();
++      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.JigsawPatternInitEvent.StructureJigsawPoolInitEvent.Village());
```
```suggestion
+      if(!JigsawPatternInitEvent.StructureJigsawPoolInitEvent.Village.isFired())
+          net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.JigsawPatternInitEvent.StructureJigsawPoolInitEvent.Village());
```

Cheaterpaul reviewed (commented):
for firing the event only once i thought about this

Cheaterpaul committed:
village fired attribute for single event firing

Cheaterpaul committed:
addendum: village fired attribute for single event firing

Cheaterpaul committed:
Merge branch '1.14.x' into 1.14_villagepools

pau101 commented:
One concern I have is the effect of changing the chance of specific types of structures. For example village houses and farms are within the same pool so if a developer wants to add a crop field of their own type that would decrease the number of houses.

The choices I see are making extensive patches to allow pools within pools, which is the preferable design or to add "categories" to entries within pools to have weights preserve likely hood of other categories, probably requiring less invasive patches.

Cheaterpaul commented:
that sounds good.

But for that a Forge Team member should state his opinion

0xilly commented:
I like the option with less invasive  patches

Cheaterpaul commented:
should the category be a Enum, IExtensibleEnum, Simple class objects or registry objects?

i assume they dont need attributes

edit:
they must be serializable. And at best compatible with new categories.

Cheaterpaul commented:
To make this possible i must edit every jigsawpiece::new call and apply a category with weight.
(with weight i mean total weight of all jigsawpieces of the same category)

This whould result in 100+ jigsawcategories (for every jigsawpattern multiple one) and modify all SingleJigsawPieces in the VillagePools (PillageOutpostPool) (PlainsVillagePools, DesertVillagePool, SavannaVillagePool, SnowyVillagePool, TaigaVillagePool, PillageOutpostPieces)

Cheaterpaul commented:
Considering this i assume the first option could be a lot better.

Or do someone has a better solution or implementation?
